src/lib.rs

```diff
@@ -95,20 +107,22 @@ impl Lottery {
     /// Arguments:
     /// * `amount`: value or tokens to participate in the lottery
     async fn enter(&mut self, amount: u128) {
-        if self.lottery_is_on() && !self.player_exists() && amount > 0 {
-            let player = Player {
-                player_id: msg::source(),
-                balance: amount,
-            };
-
-            if self.token_address.is_some() {
-                self.transfer_tokens(&msg::source(), &exec::program_id(), amount)
+        if self.lottery_is_on()
+            && !self.player_exists()
+            && amount == self.lottery_state.participation_cost
+        {
+            if let Some(ref token_address) = self.token_address {
+                Self::transfer_tokens(token_address, &msg::source(), &exec::program_id(), amount)
                     .await;
 
-                self.lottery_balance += amount;
                 debug!("Add in Fungible Token: {}", amount);
             }
 
+            let player = Player {
+                player_id: msg::source(),
+                balance: amount,
+            };
+
             let player_index = self.players.len() as u32;
             self.players.insert(player_index, player);
             msg::reply(LtEvent::PlayerAdded(player_index), 0).unwrap();
```

Comment:
`unwrap()` must be replaced with `expect()`

Reply:
Fixed